source/util/parse_number.h

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2015-2016 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_UTIL_PARSE_NUMBER_H_
+#define LIBSPIRV_UTIL_PARSE_NUMBER_H_
+
+#include <functional>
+#include <iomanip>
+#include <sstream>
+#include <tuple>
+
+#include "util/hex_float.h"
+
+namespace spvutils {
+
+// This is a lattice for tracking types.
+enum class IdTypeClass {
+  kBottom = 0,  // We have no information yet.
+  kScalarIntegerType,
+  kScalarFloatType,
+  kOtherType
+};
+
+// Contains ID type information that needs to be tracked across all Ids.
+// Bitwidth is only valid when type_class is kScalarIntegerType or
+// kScalarFloatType.
+struct IdType {
+  uint32_t bitwidth;  // Safe to assume that we will not have > 2^32 bits.
+  bool isSigned;      // This is only significant if type_class is integral.
+  IdTypeClass type_class;
+};
+
+// Default equality operator for IdType. Tests if all members are the same.
+inline bool operator==(const IdType& first, const IdType& second) {
+  return (first.bitwidth == second.bitwidth) &&
+         (first.isSigned == second.isSigned) &&
+         (first.type_class == second.type_class);
+}
+
+// Tests whether any member of the IdTypes do not match.
+inline bool operator!=(const IdType& first, const IdType& second) {
+  return !(first == second);
+}
+
+// Returns true if the type is a scalar integer type.
+inline bool isScalarIntegral(const IdType& type) {
+  return type.type_class == IdTypeClass::kScalarIntegerType;
+}
+
+// Returns true if the type is a scalar floating point type.
+inline bool isScalarFloating(const IdType& type) {
+  return type.type_class == IdTypeClass::kScalarFloatType;
+}
+
+// Returns the number of bits in the type.
+// This is only valid for bottom, scalar integer, and scalar floating
+// classes.  For bottom, assume 32 bits.
+inline int assumedBitWidth(const IdType& type) {
+  switch (type.type_class) {
+    case IdTypeClass::kBottom:
+      return 32;
+    case IdTypeClass::kScalarIntegerType:
+    case IdTypeClass::kScalarFloatType:
+      return type.bitwidth;
+    default:
+      break;
+  }
+  // We don't care about this case.
+  return 0;
+}
+
+enum class EncodeNumberResult {
```

Comment:
Oh, wait, this is actually used to mean the status of parsing? So maybe it's better to name it as `NumberParsingStatus`?